### PR TITLE
MLDB-1277 pooling performance

### PR DESCRIPTION
**Pull request by jeremybarnes**
This series of commits massively improves the performance of the pooling operations, with the MLDB-1277 test going from ~10 seconds for 1000 applications ~3 seconds for 100,000 applications.

It would be more instructive to review the commits individually than the whole lot, as they are fairly self-contained but the combination is a little unwieldy.

### CODE REVIEW

File: plugins/pooling_function.cc [hunk 1]
```diff
@@ -51,7 +52,10 @@ PoolingFunction(MldbServer * owner,
 {
     functionConfig = config.params.convert<PoolingFunctionConfig>();
 
-    set<string> validAggs = {"avg", "min", "max", "sum"};
+    SqlQueryFunctionConfig fnConfig;
+    fnConfig.query.stm.reset(new SelectStatement());
+
+    set<Utf8String> validAggs = {"avg", "min", "max", "sum"};
```
Comment (steadtler) [inline plugins/pooling_function.cc]: Why not use the official set of registered builtin aggregators?

File: plugins/pooling_function.cc [hunk 2]
```diff
@@ -60,21 +64,32 @@ PoolingFunction(MldbServer * owner,
...
             select_expr += ",";
 
-        select_expr += ML::format("%s({*}) as %s", agg, agg);
+        select_expr += ML::format("%s({*}) as %s", agg.rawData(), agg.rawData());
     }
-    select = SelectExpression::parseList(select_expr);
+    fnConfig.query.stm->select = SelectExpression::parseList(select_expr);
+    fnConfig.query.stm->from = functionConfig.embeddingDataset;
+    fnConfig.query.stm->where = SqlExpression::parse("rowName() IN (KEYS OF $words)");
+    // Force group by, since that query executor doesn't know how to determine
```
Comment (steadtler) [inline plugins/pooling_function.cc]: Should we make a story for that?

File: sql/builtin_aggregators.cc [hunk 3]
```diff
@@ -117,41 +117,27 @@ struct AggregatorT {
...
         }
+        
+        std::unordered_map<ColumnName, State> columns;
 
         void process(const ExpressionValue * args, size_t nargs)
         {
             ExcAssertEqual(nargs, 1);
             const ExpressionValue & val = args[0];
 
+            if (false /* dense */) {
```
Comment (steadtler) [inline sql/builtin_aggregators.cc]: We probably dont need to keep that?

**Comment by steadtler** [comment]
Just a couple of questions, +1 after answering

File: container_files/assets/www/doc/builtin/sql/ValueExpression.md [hunk 4]
```diff
@@ -216,11 +216,42 @@ intended. See also [the MLDB Type System](TypeSystem.md).
...
+2.  As an explicit tuple (`x IN (val1, val2, ...)`)
+3.  As the keys of a row expression (`x IN (KEYS OF expr)`)
+4.  As the values of a row expression (`x IN (VALUES OF expr)`)
+
+The first two are standard SQL; the second two are MLDB extensions and are
+made possible by MLDB's sparse data model.
+
+#### IN expression with sub-select
+
+The right hand side can also be the result of a sub `SELECT` statement.
```
Comment (guyd) [inline container_files/assets/www/doc/builtin/sql/ValueExpression.md]: I would remove this line.

**Comment by guyd** [comment]
Briefly reviewed.  +1 once the tests are passing.

File: plugins/pooling_function.cc (hunk 1 above)
Comment (jeremybarnes) [inline plugins/pooling_function.cc]: Ask @mailletf.  I simply ported the existing functionality to a new execution engine.

File: plugins/pooling_function.cc (hunk 2 above)
Comment (jeremybarnes) [inline plugins/pooling_function.cc]: Yes, yes we should. https://datacratic.atlassian.net/browse/MLDB-1300

File: plugins/pooling_function.cc (hunk 1 above)
Comment (nicolaskruchten) [inline plugins/pooling_function.cc]: These are the only aggregators we have which which make sense in the context... `count` and `pivot` don't make sense in this context and we don't have `stdev` :)

File: sql/builtin_aggregators.cc (hunk 3 above)
Comment (jeremybarnes) [inline sql/builtin_aggregators.cc]: removed

File: container_files/assets/www/doc/builtin/sql/ValueExpression.md (hunk 4 above)
Comment (jeremybarnes) [inline container_files/assets/www/doc/builtin/sql/ValueExpression.md]: Reworded

**Pull request merged**